department-of-veterans-affairs/abd-vro
# MCP-2123 - Basic sanitization of Request Body JSON input

## What was the problem?

We weren't sanitizing any of our JSON Request Body input data.

Associated tickets or Slack threads:

- [MCP-2123](https://amida.atlassian.net/browse/MCP-2123)

## How does this fix it?[^1]

Now we have an explicit place for JSON Request Body sanitation via `DisallowedPatternAdvice` which implements `RequestBodyAdvice` and pointcuts across every single occurrence of `RequestBody`. This implements a very rudimentary sanitation of the following:

- non-printable characters
- NUL/null characters

## How to test this PR
UNHAPPY PATH TO TRIGGER THE EXCEPTION:
- Step 1: `POST` to `http://localhost:8080/v2/examOrderingStatus` with the following RequestBody:
```JSON
{
  "collectionId": 999,
  "collectionStatus": "http://localhost:8080/v1/fetch-claims/%00%255c%252e%252e%255c/%252e%252e%255c/%252e%252e%255c/%252e%252e%255c/%252e%252e%255c/windows/system.ini\b\u007F\u0081\u0088/%00",
  "examOrderDateTime": "2018-11-04T17:45:61Z",
}
```
- Step 2: Verify a `400 Bad Request` response from the server with the following in the application log:
```
2023-01-09 01:15:00 2023-01-09 06:15:00.029 ERROR 1 --- [nio-8080-exec-9] g.v.v.c.e.GlobalExceptionHandler         : Unallowed patterns were found in the Request Body.
2023-01-09 01:15:00 
2023-01-09 01:15:00 gov.va.vro.controller.exception.DisallowedPatternException: Unallowed patterns were found in the Request Body. Please sanitize your Request Body input data and try again.
```
HAPPY PATH:
- Step 1: `POST` to `http://localhost:8080/v2/examOrderingStatus` with the following RequestBody:
```JSON
{
  "collectionId": 999,
  "collectionStatus": "DRAFT",
  "examOrderDateTime": "2018-11-04T17:45:61Z"
}
```
- Step 2: Verify a `200 OK` response from the server

## NOTES:
This is NOT a comprehensive sanitation filter. This only 1) adds a place to put sanitation logic inside and 2) includes only the most basic sanitizing filtration on our JSON Request Body strings. A security specialist might have to perform an audit and make recommendations for us to add further scenario logic.

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
